src/stm32f1/pwm.cpp

```diff
@@ -0,0 +1,230 @@
+#include <libhal-util/bit.hpp>
+
+namespace hal::stm32f1 {
+namespace {
+[[nodiscard]] pwm_reg_t* get_pwm_reg(peripheral p_id)
+{
+  if (p_id == peripheral::timer1) {
+    return pwm_timer1;
+  }
+  return pwm_timer8;
+}
+[[nodiscard]] peripheral get_peripheral_id(pwm_pins p_pin)
+{
+  if (p_pin == pwm_pins::pa8 || p_pin == pwm_pins::pa9 ||
+      p_pin == pwm_pins::pa10 || p_pin == pwm_pins::pa11) {
+
+    return peripheral::timer1;
+  } else {
+    hal::safe_throw(hal::operation_not_supported(nullptr));
+  }
+  return peripheral::timer8;
+}
+
+[[nodiscard]] float get_duty_cycle(pwm_pins p_pin,
+                                   uint32_t volatile* compare_register)
+{
+  // ccr / arr
+  auto peripheral_id = get_peripheral_id(p_pin);
+  pwm_reg_t* p_reg = get_pwm_reg(peripheral_id);
+
+  static constexpr auto first_nonreserved_half = bit_mask::from<0, 15>();
+
+  std::uint16_t compare_value =
+    hal::bit_extract<first_nonreserved_half>(*compare_register);
+
+  std::uint16_t arr_value =
+    hal::bit_extract<first_nonreserved_half>(p_reg->auto_reload_register);
+
+  return (static_cast<float>(compare_value) / static_cast<float>(arr_value));
+}
+void setup_channel(pwm_reg_t* p_reg, uint8_t p_channel)
+{
+
+  // uint32_t start_pos = 4 * (p_channel - 1);
+  static constexpr auto main_output_enable = bit_mask::from<15>();
+  static constexpr auto ossr = bit_mask::from<11>();
+
+  switch (p_channel) {
+    case 1:
+      static constexpr auto ccer_polarity_enable_mask = bit_mask::from(0, 1);
+      bit_modify(p_reg->cc_enable_register)
+        .insert<ccer_polarity_enable_mask>(0b01U);
+      break;
+
+    case 2:
+      static constexpr auto ccer_polarity_enable_mask_2 = bit_mask::from(4, 5);
+      bit_modify(p_reg->cc_enable_register)
+        .insert<ccer_polarity_enable_mask_2>(0b01U);
+
+      break;
+    case 3:
+      static constexpr auto ccer_polarity_enable_mask_3 = bit_mask::from(8, 9);
+      bit_modify(p_reg->cc_enable_register)
+        .insert<ccer_polarity_enable_mask_3>(0b01U);
+      break;
+    case 4:
+      static constexpr auto ccer_polarity_enable_mask_4 =
+        bit_mask::from(12, 13);
+      bit_modify(p_reg->cc_enable_register)
+        .insert<ccer_polarity_enable_mask_4>(0b01U);
+      break;
+    default:
+      break;
+  }
+
+  bit_modify(p_reg->break_and_deadtime_register)
+    .clear(ossr)
+    .set(main_output_enable);  // complementary channel stuff
+
+  return;
+}
+void setup(pwm_pins p_pin)
+{
+
+  static constexpr auto clock_division = bit_mask::from<8, 9>();
+  static constexpr auto edge_aligned_mode = bit_mask::from<5, 6>();  // set to
+                                                                     // 00
+  static constexpr auto direction = bit_mask::from<4>();
+
+  static constexpr auto output_compare_odd = bit_mask::from<4, 6>();
+  static constexpr auto output_compare_even = bit_mask::from<12, 14>();
+  static constexpr auto counter_enable = bit_mask::from<0>();  // set this to 1
```

Comment:
```suggestion
  static constexpr auto counter_enable = bit_mask::from<0>();
```
Redundant comment. It should be obvious that the counter enable when set to a 1 enables the counter. If it needed to be set to 0 then that would be non-obvious.